### ISSUE TITLE
Compiling for Android: is it possible to use the 'userelocatableinc' option?

### ISSUE THREAD

**Issue by AdamDotNet** [arsv/perl-cross#53]
Hello.

I've been spending quite some time working with perl-cross to try and get perl compiled for Android 8. Prior to my coworker's departure from the company, he would use the CCTools app on a device or emulator and run configure/make right on the device/emulator. I don't think CCTools supports Android 8 yet, nor did Google release an ARM emulator for Android 8, as of this writing at least.

Anyway, inspecting a Config.pm from a working build, for example Android 6 x86, it has the following:

``` perl
sub relocate_inc { ... }
...
tie %Config, 'Config', {
    archlibexp => relocate_inc('.../../lib/5.22.1/i686-linux-android'),
   ...
}
```

This lets us run perl from within our app without regard to where in our app we stash perl. My understanding is that if you call perl's `Configure` with the option `-Duserelocatableinc` the result should be similar to the above. However, my resulting Config.pm file does not have the relocate_inc method and hard codes paths based on `--prefix` or the default prefix.

Here's what I am trying:
1. Build environment: Ubuntu 16 x64 (technically ran through Bash on Ubuntu on Windows 10)
2. Android NDK is extracted at /opt/android-ndk (r15c)
3. Build directories: /data/local/perl (for the result) and /data/local/perl-5.22.4 (for the source code)
4. perl-cross 1.1.6 unzipped onto perl source 5.22.4
5. Environment variables, for x86 for example
``` bash
export ANDROID=/opt/android-ndk
export TOOLCHAIN=x86-4.9/prebuilt/linux-x86_64
export PLATFORM=$ANDROID/platforms/android-26/arch-x86
export PATH=$PATH:$ANDROID/toolchains/$TOOLCHAIN/bin
```
6. Run Configure
`./Configure -Duserelocatableinc --target=i686-linux-android --sysroot=$PLATFORM --all-static -Accflags="-pie -fPIE" -Aldflags="-pie -fPIE"`
7. `make`, etc.

At some point prior to discovering `userelocatableinc`, I was setting a prefix to `/data/local/perl` and that worked on the device, _if I stashed perl to /data/local/perl_. Otherwise no packages are found and it prints the `@INC` array and kills the script. Regardless of setting a prefix or not though, when `userelocatableinc` is set, the resulting Config.pm file always has hard coded paths, and does not include the `relocate_inc` method.

I wanted to make builds comparable to the older, working builds made with CCTools, but does perl-cross support `userelocatableinc`? Is the only solution to prefix the build to where we ultimately stash perl inside our app and hope we never choose to move it elsewhere? E.g. `/data/data/com.app.name/perl`. I'd feel a lot safer if I could make path independent builds.

**Comment by arsv** [comment]
No proper support at this moment, I'll look into adding it. In the meantime, this trick should work:

    ./configure -Dprefix=.../.. -Dbin=.../
    make perl

The resulting executable will look for modules in ../lib relative to its own location.

`make install` will not work with that configuration however since it cannot handle .../ in installation paths. A simple workaround: try building everything with a fixed prefix (any prefix, like /usr/local) to get the modules, install it into a temp directory, then re-configure with prefix=.../.. to build relocatable `perl` executable, and use it with modules from the complete build.

**Comment by AdamDotNet** [comment]
@arsv Thanks for your quick reply!

With all the settings mentioned in my original post, and adding your suggesting, my complete configure arguments are as follows:

`./Configure -Duserelocatableinc -Dprefix=.../.. -Dbin=.../ --target=i686-linux-android --sysroot=$PLATFORM --all-static -Accflags="-pie -fPIE" -Aldflags="-pie -fPIE"`

Config.pm had the expected `sub relocate_inc` and of course the paths had their .../.. prefixes.

I threw caution to the wind and ran `make DESTDIR=/data/local/perl install.perl` even though the prefix was .../.. and it truly did weird things, but mostly worked! For example, the perl binary was in the source directory because make tried to stick it in `...//perl` which ended up just sticking it in the current directory, but everything else was in bin and lib as expected. It also ended up creating some weird recursive directory structure when it created a directory called perl... (the contents were the directory listing of its parent directory) but if you don't mind nuking your entire build working directory, no problem there either.

Furthermore, testing it out on the emulator, perl -V happily found the right `@INC` paths even if I moved it around.

Finally, integrating it into my app, it all worked out! You are a life saver!

If you end up adding direct support for `userelocatableinc` without using the prefix workaround, I'd be willing to try out the solution to help you validate it.

**Comment by arsv** [comment]
Ok, -Duserelocatableinc should work right away now. Please test it with your setup if possible, and if everything works well I'll tag the next version.

**Comment by AdamDotNet** [comment]
I was unable to run the latest code on master. I tried downloading in various ways:
1. On Windows, simply download as zip file, extract over perl source and then run on Ubuntu 16.
2. On Windows, use git clone, copy to perl source, then run on Ubuntu.
3. On Ubuntu, use git clone, copy to perl source, then run on Ubuntu.

I always got the same error at the make stage:

```
patch -p1 -i cnf/diffs/perl5-5.22.4/configpm.patch && touch cnf/diffs/perl5-5.22.4/configpm.applied
patch unexpectedly ends in middle of line
patch: **** Only garbage was found in the patch input.
Makefile:67: recipe for target 'cnf/diffs/perl5-5.22.4/configpm.applied' failed
make: *** [cnf/diffs/perl5-5.22.4/configpm.applied] Error 2
```

What's the right way to try out your latest code?

**Comment by arsv** [comment]
git clone should work, strange. I see nothing wrong with that file, and it wasn't even changed since the last version. However, it's a symlink. Chances are your git cannot handle symlinks for some reason.

Can you try this archive: https://github.com/arsv/perl-cross/raw/temp/perl-cross-1.1.7-pre.tar.gz

zip archives won't work most likely, I don't think zip format has any support for symlinks.

**Comment by AdamDotNet** [comment]
You're right. When I used the zip download and even git clone, the patches were actual files whose contents were the relative paths to the other patches.

cnf/diffs/perl5-5.22.4/configpm.patch contents were
`../perl5-5.22.3/configpm.patch`

When I extracted the tar you gave me, the symlinks were intact.

In any case, the trial was a success! Configure was as follows:
`./Configure -Duserelocatableinc --target=i686-linux-android --sysroot=$PLATFORM --all-static -Accflags="-pie -fPIE" -Aldflags="-pie -fPIE"`

The Config.pm file had the .../.. paths and the sub to figure them out and it indeed ran in my app correctly. Consider it a great success! Well done.
